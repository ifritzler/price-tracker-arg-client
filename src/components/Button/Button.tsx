import Link from "next/link";
import { ButtonHTMLAttributes } from "react";

type Props = {
  text: string;
  typeb: "button" | "submit" | undefined;
};

type PropsLink = {
  text: string;
  typeb: "button-link";
  url: string;
  icon?: JSX.Element;
  onlyIcon?: boolean;
};

export default function Button(
  props: (PropsLink | Props) & ButtonHTMLAttributes<HTMLButtonElement>
) {
  const { text, typeb, className, ...rest } = props;

  if (typeb === "button-link") {
    const { url, icon, onlyIcon = false, ...rest } = props;

    return (
      <button
        className={`rounded-md ${onlyIcon ? "p-2" : "px-4 py-2"} flex items-center justify-center bg-[--color-primary] text-white w-full shadow-sm ${className}`}
        type={"button"}
      >
        <Link href={url || "/"} className="flex flex-row gap-1">
          {!onlyIcon && <span>{text}</span>}
          {icon && <i>{icon}</i>}
        </Link>
      </button>
    );
  }

  return (
    <button
      className={`rounded-[20px] px-4 py-2 flex place-content-center bg-[--color-primary] shadow-sm w-full text-white ${className}`}
      type={typeb}
      {...rest}
    >
      {text}
    </button>
  );
}
