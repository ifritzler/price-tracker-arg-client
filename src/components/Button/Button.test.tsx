import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Button from "./Button";

describe("Button", () => {
    it("check renders button unchanged", () => {
        const { container } = render(<Button text="Button" typeb="button"/>);
        expect(container).toMatchSnapshot("button");
        
        const { container: containerLinkButton } = render(<Button text="Button" typeb="button-link" url="www.google.com"/>);
        expect(containerLinkButton).toMatchSnapshot("link-button");
    })
    
    it("text button should render", () => {
        render(<Button text="Button 1" typeb="button"/>);
        const button = screen.getByRole('button')

        expect(button).toHaveTextContent("Button 1")
    })
    
    it("button should be a link if url exists", () => {
        render(<Button text="Button 1" typeb="button-link" url="www.google.com"/>);
        
        const button = screen.getByRole('link')
    
        expect(button).toHaveAttribute('href', 'www.google.com')
    })

    it("url must be root if url is empty", () => {
        render(<Button text="Button 1" typeb="button-link" url=""/>);
        
        const button = screen.getByRole('link')
    
        expect(button).toHaveAttribute('href', '/')
    })

    it("text must shows when onlyIcon is not defined", () => {
        render(<Button text="Button 1" typeb="button-link" url="www.google.com"/>);
        
        const button = screen.getByRole('link')
        
        expect(button).toHaveTextContent('Button 1')
    })
    
    it("Behaviour button on click", () => {
        const onClickMock = jest.fn();
        render(<Button text="Button 1" typeb="button" onClick={onClickMock}/>);
        
        const button = screen.getByRole('button')
        fireEvent.click(button)

        expect(onClickMock).toHaveBeenCalled();
        expect(onClickMock).toHaveBeenCalledTimes(1);
    })
})