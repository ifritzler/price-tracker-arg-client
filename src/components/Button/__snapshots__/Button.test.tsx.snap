// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button check renders button unchanged: button 1`] = `
<div>
  <button
    class="rounded-[20px] px-4 py-2 flex place-content-center bg-[--color-primary] shadow-sm w-full text-white undefined"
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button check renders button unchanged: link-button 1`] = `
<div>
  <button
    class="rounded-md px-4 py-2 flex items-center justify-center bg-[--color-primary] text-white w-full shadow-sm undefined"
    type="button"
  >
    <a
      class="flex flex-row gap-1"
      href="www.google.com"
    >
      <span>
        Button
      </span>
    </a>
  </button>
</div>
`;
